In [1]:
with open('data/day_2.txt') as f:
    for line in f:
        program_str = line

print(program_str)

1,0,0,3,1,1,2,3,1,3,4,3,1,5,0,3,2,9,1,19,1,19,6,23,2,6,23,27,2,27,9,31,1,5,31,35,1,35,10,39,2,39,9,43,1,5,43,47,2,47,10,51,1,51,6,55,1,5,55,59,2,6,59,63,2,63,6,67,1,5,67,71,1,71,9,75,2,75,10,79,1,79,5,83,1,10,83,87,1,5,87,91,2,13,91,95,1,95,10,99,2,99,13,103,1,103,5,107,1,107,13,111,2,111,9,115,1,6,115,119,2,119,6,123,1,123,6,127,1,127,9,131,1,6,131,135,1,135,2,139,1,139,10,0,99,2,0,14,0


In [2]:
program_split = program_str.split(',')

program = [int(value) for value in program_split]

print(program)

[1, 0, 0, 3, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 9, 1, 19, 1, 19, 6, 23, 2, 6, 23, 27, 2, 27, 9, 31, 1, 5, 31, 35, 1, 35, 10, 39, 2, 39, 9, 43, 1, 5, 43, 47, 2, 47, 10, 51, 1, 51, 6, 55, 1, 5, 55, 59, 2, 6, 59, 63, 2, 63, 6, 67, 1, 5, 67, 71, 1, 71, 9, 75, 2, 75, 10, 79, 1, 79, 5, 83, 1, 10, 83, 87, 1, 5, 87, 91, 2, 13, 91, 95, 1, 95, 10, 99, 2, 99, 13, 103, 1, 103, 5, 107, 1, 107, 13, 111, 2, 111, 9, 115, 1, 6, 115, 119, 2, 119, 6, 123, 1, 123, 6, 127, 1, 127, 9, 131, 1, 6, 131, 135, 1, 135, 2, 139, 1, 139, 10, 0, 99, 2, 0, 14, 0]


Restore the program to its state before the crash:

In [3]:
program[1] = 12
program[2] = 2

print(program)

[1, 12, 2, 3, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 9, 1, 19, 1, 19, 6, 23, 2, 6, 23, 27, 2, 27, 9, 31, 1, 5, 31, 35, 1, 35, 10, 39, 2, 39, 9, 43, 1, 5, 43, 47, 2, 47, 10, 51, 1, 51, 6, 55, 1, 5, 55, 59, 2, 6, 59, 63, 2, 63, 6, 67, 1, 5, 67, 71, 1, 71, 9, 75, 2, 75, 10, 79, 1, 79, 5, 83, 1, 10, 83, 87, 1, 5, 87, 91, 2, 13, 91, 95, 1, 95, 10, 99, 2, 99, 13, 103, 1, 103, 5, 107, 1, 107, 13, 111, 2, 111, 9, 115, 1, 6, 115, 119, 2, 119, 6, 123, 1, 123, 6, 127, 1, 127, 9, 131, 1, 6, 131, 135, 1, 135, 2, 139, 1, 139, 10, 0, 99, 2, 0, 14, 0]


What value is left in position 0 after the program halts?

In [102]:
def run_opcode_beta0(program):
    
    index = range(0, len(program))
    
    for i in index:
        if program[i] == 1:
            program[program[i + 3]] = program[i + 1] + program[i + 2]
        elif program[i] == 2:
            program[program[i + 3]] = program[i + 1] * program[i + 2]
        elif program[i] == 99:
            break
        else:
            continue
    
    return program

# TO DO: Make it skip ahead 4 after a 1 or 2

Build up to function:

In [7]:
# First, see if the indexing works

def run_opcode_beta1(program):
    
    index = range(len(program))
    
    for i in index: 
        program[i] = index[i]
    
    return program

Run some tests:

In [8]:
test1 = [1, 0, 0, 0, 99]
print(run_opcode_beta1(test1))

[0, 1, 2, 3, 4]


In [12]:
def run_opcode_beta2(program):
    
    index = range(len(program))
    
    for i in index: 
        if program[i] == 1:
            program[program[i + 3]] = program[program[i + 1]] + program[program[i + 2]]
        elif program[i] == 2: 
            program[program[i + 3]] = program[program[i + 1]] * program[program[i + 2]]
        elif program[i] == 99:
            break
        else: 
            continue
    
    return program

In [13]:
test1 = [1, 0, 0, 0, 99]
print(run_opcode_beta2(test1))

[2, 0, 0, 0, 99]


Ah I see - the last 4 positions won't run because they're preceded by 99. Looking before, the last opcode was to add the values in position 139 and 10 and store them in position 0. 

In [14]:
test2 = [2,3,0,3,99]
print(run_opcode_beta2(test2))

[2, 3, 0, 6, 99]


In [16]:
test3 = [2,4,4,5,99,0]
print(run_opcode_beta2(test3))

[2, 4, 4, 5, 99, 9801]


In [19]:
test4 = [1, 1, 1, 4, 99, 5, 6, 0, 99]
print(run_opcode_beta2(test4))

[30, 1, 3, 4, 2, 5, 6, 0, 99]


The problem is that it needs to skip ahead 4 positions after each operator:

In [22]:
def run_opcode_beta3(program):
    
    position = 0
    
    while position < len(program): 
        if program[position] == 1:
            program[program[position + 3]] = program[program[position + 1]] + program[program[position + 2]]
            position += 4
        if program[position] == 2:
            program[program[position + 3]] = program[program[position + 1]] * program[program[position + 2]]
            position += 4
        if program[position] == 99:
            break
        
    return program

In [23]:
test1 = [1, 0, 0, 0, 99]
print(run_opcode_beta3(test1))

[2, 0, 0, 0, 99]


In [24]:
test2 = [2,3,0,3,99]
print(run_opcode_beta3(test2))

[2, 3, 0, 6, 99]


In [25]:
test3 = [2,4,4,5,99,0]
print(run_opcode_beta3(test3))

[2, 4, 4, 5, 99, 9801]


In [26]:
test4 = [1, 1, 1, 4, 99, 5, 6, 0, 99]
print(run_opcode_beta3(test4))

[30, 1, 1, 4, 2, 5, 6, 0, 99]


So we're going with this version:

In [4]:
def run_opcode(program):
    
    pointer = 0
    
    while pointer < len(program): 
        if program[pointer] == 1:
            program[program[pointer + 3]] = program[program[pointer + 1]] + program[program[pointer + 2]]
            pointer += 4
        if program[pointer] == 2:
            program[program[pointer + 3]] = program[program[pointer + 1]] * program[program[pointer + 2]]
            pointer += 4
        if program[pointer] == 99:
            break
        
    return program

In [5]:
print(run_opcode(program))

[6730673, 12, 2, 2, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 9, 1, 36, 1, 19, 6, 38, 2, 6, 23, 76, 2, 27, 9, 228, 1, 5, 31, 229, 1, 35, 10, 233, 2, 39, 9, 699, 1, 5, 43, 700, 2, 47, 10, 2800, 1, 51, 6, 2802, 1, 5, 55, 2803, 2, 6, 59, 5606, 2, 63, 6, 11212, 1, 5, 67, 11213, 1, 71, 9, 11216, 2, 75, 10, 44864, 1, 79, 5, 44865, 1, 10, 83, 44869, 1, 5, 87, 44870, 2, 13, 91, 224350, 1, 95, 10, 224354, 2, 99, 13, 1121770, 1, 103, 5, 1121771, 1, 107, 13, 1121776, 2, 111, 9, 3365328, 1, 6, 115, 3365330, 2, 119, 6, 6730660, 1, 123, 6, 6730662, 1, 127, 9, 6730665, 1, 6, 131, 6730667, 1, 135, 2, 6730669, 1, 139, 10, 0, 99, 2, 0, 14, 0]


Answer: The value in position 0 is 6730673

## Part 2

In [4]:
print(program) # In its state before the crash. 

# Position 0 = Output
# Position 1 = Noun (Input)
# Position 2 = Verb (Input)

[1, 12, 2, 3, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 9, 1, 19, 1, 19, 6, 23, 2, 6, 23, 27, 2, 27, 9, 31, 1, 5, 31, 35, 1, 35, 10, 39, 2, 39, 9, 43, 1, 5, 43, 47, 2, 47, 10, 51, 1, 51, 6, 55, 1, 5, 55, 59, 2, 6, 59, 63, 2, 63, 6, 67, 1, 5, 67, 71, 1, 71, 9, 75, 2, 75, 10, 79, 1, 79, 5, 83, 1, 10, 83, 87, 1, 5, 87, 91, 2, 13, 91, 95, 1, 95, 10, 99, 2, 99, 13, 103, 1, 103, 5, 107, 1, 107, 13, 111, 2, 111, 9, 115, 1, 6, 115, 119, 2, 119, 6, 123, 1, 123, 6, 127, 1, 127, 9, 131, 1, 6, 131, 135, 1, 135, 2, 139, 1, 139, 10, 0, 99, 2, 0, 14, 0]


**Instructions:** Find the input noun and verb that cause the program to produce the output 19690720. What is 100 * noun + verb? (For example, if noun=12 and verb=2, the answer would be 1202.)

Try brute force method, trying values from 0 to 10

In [6]:
def test_output(program, input1, input2):
    
    program[1] = input1
    program[2] = input2
    
    new = run_opcode(program)
    output = new[0]
    return output

In [ ]:
print(test_output(program, 1, 3))

In [ ]:
print("hello")